upstairs/src/lib.rs

```diff
@@ -8676,16 +8677,22 @@ impl GtoS {
      */
     #[instrument]
     async fn transfer(&mut self) {
-        if let Some(guest_buffer) = &mut self.guest_buffer {
-            self.completed.sort_unstable();
-            assert!(!self.completed.is_empty());
+        assert!(!self.completed.is_empty());
 
-            let mut offset = 0;
-            let mut vec = guest_buffer.as_vec().await;
-            let mut owned_vec = guest_buffer.owned_vec().await;
+        for ds_id in &self.completed {
+            if let Some(guest_buffer) = self.guest_buffers.get_mut(ds_id) {
+                let mut offset = 0;
+                let mut vec = guest_buffer.as_vec().await;
+                let mut owned_vec = guest_buffer.owned_vec().await;
 
-            for ds_id in self.completed.iter() {
-                let responses = self.downstairs_buffer.remove(ds_id).unwrap();
+                /*
+                 * Should this panic?  If the caller is requesting a transfer,
+                 * the guest_buffer should exist. If it does not exist, then
+                 * either there is a real problem, or the operation was a write
+                 * or flush and why are we requesting a transfer for those.
+                 */
+                let responses =
+                    self.downstairs_responses.remove(ds_id).unwrap();
```

Comment:
It makes me nervous that this function leaves the object in an invalid state; if someone managed to call `transfer` twice, it would fail.

Looking below, we call `transfer` in a specific context:
```rust
                if result.is_ok() && !gtos_job.guest_buffers.is_empty() {
                    gtos_job.transfer().await;
                }

                gtos_job.notify(result);
```
Could we instead have something like `gtos.finalize(result)`, which does the same combination of `transfer` and `notify` but consumes `mut self` as it does?

Reply:
agreed, added in 8fa5a530c3fc2389ac51ff7e985c44afb5995af7